data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamAcknowledgementManager.java

```diff
@@ -0,0 +1,127 @@
+        this.acknowledgementSetManager = acknowledgementSetManager;
+        this.partitionCheckpoint = partitionCheckpoint;
+        this.partitionAcknowledgmentTimeout = partitionAcknowledgmentTimeout;
+        this.acknowledgementMonitorWaitTimeInMs = acknowledgementMonitorWaitTimeInMs;
+        this.checkPointIntervalInMs = checkPointIntervalInMs;
+        executorService = Executors.newSingleThreadExecutor(BackgroundThreadFactory.defaultExecutorThreadFactory("mongodb-stream-ack-monitor"));
+    }
+
+    void init(final Consumer<Void> stopWorkerConsumer) {
+        enableAcknowledgement = true;
+        executorService.submit(() -> monitorAcknowledgment(executorService, stopWorkerConsumer));
+    }
+
+    private void monitorAcknowledgment(final ExecutorService executorService, final Consumer<Void> stopWorkerConsumer) {
+        long lastCheckpointTime = System.currentTimeMillis();
+        CheckpointStatus lastCheckpointStatus = null;
+        while (!Thread.currentThread().isInterrupted()) {
+            final CheckpointStatus checkpointStatus = checkpoints.peek();
+            if (checkpointStatus != null) {
+                if (checkpointStatus.isAcknowledged()) {
+                    lastCheckpointStatus = checkpoints.poll();
+                    ackStatus.remove(checkpointStatus.getResumeToken());
+                    if (System.currentTimeMillis() - lastCheckpointTime >= checkPointIntervalInMs) {
+                        LOG.debug("Perform regular checkpointing for resume token {} at record count {}", checkpointStatus.getResumeToken(), checkpointStatus.getRecordCount());
+                        partitionCheckpoint.checkpoint(checkpointStatus.getResumeToken(), checkpointStatus.getRecordCount());
+                        lastCheckpointTime = System.currentTimeMillis();
+                    }
+                } else {
+                    LOG.debug("Checkpoint not complete for resume token {}", checkpointStatus.getResumeToken());
+                    final Duration ackWaitDuration = Duration.between(Instant.ofEpochMilli(checkpointStatus.getCreateTimestamp()), Instant.now());
+                    // Acknowledgement not received for the checkpoint after twice ack wait time
+                    if (ackWaitDuration.getSeconds() >= partitionAcknowledgmentTimeout.getSeconds() * 2) {
+                        // Give up partition and should interrupt parent thread to stop processing stream
+                        if (lastCheckpointStatus != null && lastCheckpointStatus.isAcknowledged()) {
+                            partitionCheckpoint.checkpoint(lastCheckpointStatus.getResumeToken(), lastCheckpointStatus.getRecordCount());
+                        }
+                        LOG.warn("Acknowledgement not received for the checkpoint {} past wait time. Giving up partition.", checkpointStatus.getResumeToken());
+                        partitionCheckpoint.giveUpPartition();
+                        break;
```

Comment:
I don't understand this. This will end the thread and then also the stream thread. But, would these ever start again? It seems that when this happens, this node is now done working and unable to contribute.

Reply:
Yes, this will end the stream. New node or same node can acquire the partition and work from the checkpoint state.

Reply:
OK. I see. The `StreamScheduler` is still running and can re-acquire.